@ operator lets the assertion rewrite hang up

Hi there,

It seems like the @ operator isn't recognized by something int the assertion_rewrite procedure.
With --assert=plain there is no problem.

I did some fake test to single out the problem in fake_matrix.py: 
```python
class FakeMatrix:

    def __init__(self, num):
        self.num = num

    def __matmul__(self, other):
        return self.num * other.num
```
and the test file: 
```python
from fake_matrix import FakeMatrix

def test_multmat_operator():
    assert FakeMatrix(2) @ FakeMatrix(3) == 6
```
traceback: 
```
platform linux -- Python 3.5.0+, pytest-2.8.5, py-1.4.31, pluggy-0.3.1 -- /home/snail/.virtualenvs/fip/bin/python3.5
cachedir: .cache
rootdir: /home/snail/comp/python/fluent/pytest_bug_report, inifile: 
collected 0 items / 1 errors 

================================================================= ERRORS =================================================================
____________________________________________________ ERROR collecting test_multmat.py ____________________________________________________
/home/snail/.virtualenvs/fip/lib/python3.5/site-packages/py/_path/local.py:650: in pyimport
    __import__(modname)
<frozen importlib._bootstrap>:969: in _find_and_load
    ???
<frozen importlib._bootstrap>:954: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:892: in _find_spec
    ???
<frozen importlib._bootstrap>:873: in _find_spec_legacy
    ???
/home/snail/.virtualenvs/fip/lib/python3.5/site-packages/_pytest/assertion/rewrite.py:143: in find_module
    source_stat, co = _rewrite_test(state, fn_pypath)
/home/snail/.virtualenvs/fip/lib/python3.5/site-packages/_pytest/assertion/rewrite.py:290: in _rewrite_test
    rewrite_asserts(tree)
/home/snail/.virtualenvs/fip/lib/python3.5/site-packages/_pytest/assertion/rewrite.py:348: in rewrite_asserts
    AssertionRewriter().run(mod)
/home/snail/.virtualenvs/fip/lib/python3.5/site-packages/_pytest/assertion/rewrite.py:569: in run
    new.extend(self.visit(child))
/usr/lib/python3.5/ast.py:245: in visit
    return visitor(node)
/home/snail/.virtualenvs/fip/lib/python3.5/site-packages/_pytest/assertion/rewrite.py:677: in visit_Assert
    top_condition, explanation = self.visit(assert_.test)
/usr/lib/python3.5/ast.py:245: in visit
    return visitor(node)
/home/snail/.virtualenvs/fip/lib/python3.5/site-packages/_pytest/assertion/rewrite.py:852: in visit_Compare
    left_res, left_expl = self.visit(comp.left)
/usr/lib/python3.5/ast.py:245: in visit
    return visitor(node)
/home/snail/.virtualenvs/fip/lib/python3.5/site-packages/_pytest/assertion/rewrite.py:761: in visit_BinOp
    symbol = binop_map[binop.op.__class__]
E   KeyError: <class '_ast.MatMult'>
```

My system is debian linux stretch, and as you can see I'm using Python 3.5.0+, pytest-2.8.5 on a virtualenv

Not strictly involved with this bug, but if you over can't sleep, you can look at pytest's generated rewritten ast, converted back to python with [these instructions](https://gist.github.com/tomviner/c3537c2f2b2b8172f83e) (now updated for Python 3).

For a failing version of your test @Shinkenjoe it generates this:
```
py.test test_matrix.py -s
============================================================== test session starts ===============================================================
platform linux -- Python 3.5.1, pytest-2.8.8.dev1, py-1.4.31, pluggy-0.3.1
rootdir: /home/tom/Dropbox/DellXPS-backup/dev/pytest-dev/ast_rewrite_gist, inifile: 
collecting 0 items

import builtins as @py_builtins
import _pytest.assertion.rewrite as @pytest_ar
from fake_matrix import FakeMatrix

def test_multmat_operator_non_equal():
    @py_assert1 = 2
    @py_assert3 = FakeMatrix(@py_assert1)
    @py_assert6 = 3
    @py_assert8 = FakeMatrix(@py_assert6)
    @py_assert10 = @py_assert3 @ @py_assert8
    @py_assert12 = 5
    @py_assert11 = @py_assert10 == @py_assert12
    if not @py_assert11:
        @py_format14 = @pytest_ar._call_reprcompare(('==',), (@py_assert11,
            ), (
            """(%(py4)s
{%(py4)s = %(py0)s(%(py2)s)
} @ %(py9)s
{%(py9)s = %(py5)s(%(py7)s)
}) == %(py13)s"""
            ,), (@py_assert10, @py_assert12)) % {'py9': @pytest_ar.
            _saferepr(@py_assert8), 'py13': @pytest_ar._saferepr(
            @py_assert12), 'py2': @pytest_ar._saferepr(@py_assert1), 'py4':
            @pytest_ar._saferepr(@py_assert3), 'py0': @pytest_ar._saferepr(
            FakeMatrix) if 'FakeMatrix' in @py_builtins.locals() or
            @pytest_ar._should_repr_global_name(FakeMatrix) else
            'FakeMatrix', 'py5': @pytest_ar._saferepr(FakeMatrix) if 
            'FakeMatrix' in @py_builtins.locals() or @pytest_ar.
            _should_repr_global_name(FakeMatrix) else 'FakeMatrix', 'py7':
            @pytest_ar._saferepr(@py_assert6)}
        @py_format16 = ('' + 'assert %(py15)s') % {'py15': @py_format14}
        raise AssertionError(@pytest_ar._format_explanation(@py_format16))
    (@py_assert1) = (@py_assert3) = (@py_assert6) = (@py_assert8) = (
        @py_assert10) = (@py_assert11) = (@py_assert12) = None

collected 1 items 

test_matrix.py F

==================================================================== FAILURES ====================================================================
________________________________________________________ test_multmat_operator_non_equal _________________________________________________________

    def test_multmat_operator_non_equal():
>       assert FakeMatrix(2) @ FakeMatrix(3) == 5
E       assert (FakeMatrix(2) @ FakeMatrix(3)) == 5
E        +  where FakeMatrix(2) = FakeMatrix(2)
E        +  and   FakeMatrix(3) = FakeMatrix(3)

test_matrix.py:4: AssertionError
============================================================ 1 failed in 0.01 seconds ============================================================
```

--- a/CHANGELOG
+++ b/CHANGELOG
@@ -1,6 +1,9 @@
 2.8.8.dev1
 ----------
 
+- fix #1290: support Python 3.5's @ operator in assertion rewriting.
+  Thanks Shinkenjoe for report with test case and Tom Viner for the PR.
+
 2.8.7
 -----
 

--- a/_pytest/assertion/rewrite.py
+++ b/_pytest/assertion/rewrite.py
@@ -453,6 +453,11 @@ binop_map = {
     ast.In: "in",
     ast.NotIn: "not in"
 }
+# Python 3.5+ compatibility
+try:
+    binop_map[ast.MatMult] = "@"
+except AttributeError:
+    pass
 
 # Python 3.4+ compatibility
 if hasattr(ast, "NameConstant"):

--- a/testing/test_assertrewrite.py
+++ b/testing/test_assertrewrite.py
@@ -279,6 +279,19 @@ class TestAssertionRewrite:
             assert False or 4 % 2
         assert getmsg(f) == "assert (False or (4 % 2))"
 
+    @pytest.mark.skipif("sys.version_info < (3,5)")
+    def test_at_operator_issue1290(self, testdir):
+        testdir.makepyfile("""
+            class Matrix:
+                def __init__(self, num):
+                    self.num = num
+                def __matmul__(self, other):
+                    return self.num * other.num
+
+            def test_multmat_operator():
+                assert Matrix(2) @ Matrix(3) == 6""")
+        testdir.runpytest().assert_outcomes(passed=1)
+
     def test_call(self):
         def g(a=42, *args, **kwargs):
             return False

